Auto copy selection to primary clipboard.

cathay4t opened the pull request:
With `save_to_clipboard: true` in `selection` section of
`alacritty.yml`. The selected string will automatically copy to primary
clipboard.

cathay4t commented:
Tested on Archlinux and OSX 10.13.6.

unphased commented:
I tested this and it dies with a crash if `save_to_clipboard` is missing in `selection`. That probably needs to be fixed, but otherwise the functionality looks good!!!

cathay4t commented:
@unphased Thanks for test. I will fix it ASAP.

cathay4t commented:
@unphased I have uploaded the new patch and it works on my Archlinux, can you try it again? Thank you.

unphased commented:
OK, so it defaults to true when the config flag is missing? I like that, but I dont know if thats what should be done. 

Looks good to me but maybe you should look into syncing with master and checking whats going on with the failing travis build.

jwilm commented on src/input.rs:
```diff
@@ -469,7 +470,7 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
         }
 
-        self.ctx.copy_selection(Buffer::Selection);
+        self.ctx.copy_selection(self.selection_buffer.clone());
```
If `Copy` is also derived on the `Buffer` enum, this `.clone()` call can be removed.

jwilm commented on src/config.rs:
```diff
@@ -40,12 +40,14 @@ fn true_bool() -> bool {
 #[derive(Clone, Debug, Deserialize)]
 pub struct Selection {
     pub semantic_escape_chars: String,
+    pub save_to_clipboard: Option<bool>,
```
It looks like there's no code path for this to be `None`; a `#[serde(default)]` attribute on this field will solve the config back-compat issue without need for option handling elsewhere in the code.

jwilm commented on alacritty.yml:
```diff
@@ -253,6 +253,10 @@ mouse:
+  # When set to true, selection string will be copy to primary clipborad.
+  # Default value is false, the selection string will be copy the selection
+  # clipborad.
+  save_to_clipboard: false
```
Please also add this to `alacritty_macos.yml`. Should the default be true on that platform?

chrisduerr replied:
To fall in line with the rest of the formatting in the `alacritty.yml`, an empty line should be added before this new block.

The default value also doesn't need to be documented when it's present in the default configuration file, I would propose rewording this to:
```yml
# When set to `true`, the selection string will be copied to the primary clipboard.
# Otherwise it will be copied to the selection clipboard.
```

This is just a proposal, if you have any better ideas please feel free to share them.

cathay4t commented:
@jwilm Thanks for the review. I have uploaded the new patch based on your suggestions.

jwilm commented:
@cathay4t it seems that the tests are broken on this branch. Would you mind checking that out?

cathay4t commented:
The clippy reports some false alarm on `Use of undeclared type or module Buffer`. Let me find a way
to ease that.

cathay4t commented:
@jwilm CI is happy now.

nixpulvis commented on src/event.rs:
```diff
@@ -496,5 +499,10 @@ impl<N: Notify> Processor<N> {
+            ::copypasta::Buffer::Primary
+        } else {
+            ::copypasta::Buffer::Selection
+        };
```
Should this really be one or the other? Should selecting things with `save_to_clipboard` on also fill the selection buffer?

chrisduerr replied:
I support the idea to still paste it into the selection clipboard.

chrisduerr commented on alacritty_macos.yml:
```diff
@@ -232,6 +232,11 @@ mouse:
+  # Default value is false, the selection string will be copy the selection
+  # clipborad.
+  save_to_clipboard: true
+
```
For the macOS documentation the same issues mentioned for the linux configuration file apply.

However since macOS does not have a selection clipboard as far as I know, the documentation should be changed to reflect that.

There's also an additional newline at the end of the block which needs to be removed.

chrisduerr commented on src/config.rs:
```diff
@@ -40,12 +40,15 @@ fn true_bool() -> bool {
 pub struct Selection {
     pub semantic_escape_chars: String,
+    #[serde(default)]
+    pub save_to_clipboard: bool,
```
Since this doesn't have a failure case implemented, it would completely invalidate the Alacritty configuration file if the user specifies a value like `trues`.

To resolve this the `deserialize_with = "failure_default"` annotation should be added. You should find the concrete way to do this with several other bindings in the `config.rs`.

chrisduerr commented on src/event.rs:
```diff
@@ -281,6 +282,7 @@ impl<N: Notify> Processor<N> {
             last_modifiers: Default::default(),
             pending_events: Vec::with_capacity(4),
             window_changes: Default::default(),
+            selection_buffer: ::copypasta::Buffer::Primary,
```
Using an absolute path inside the code like this is usually not optimal. Since the buffer is used multiple times in this file, I propose importing it at the beginning of it. If the concern was that `Buffer` doesn't really say much about what it actually is (which would be understandable), I'd propose just importing it as `ClipboardBuffer`.

chrisduerr commented on src/input.rs:
```diff
@@ -24,7 +24,8 @@ use std::process::Command;
 
-use copypasta::{Clipboard, Load, Buffer};
+use copypasta::{Clipboard, Load};
+use copypasta::Buffer as ClipboardBuffer;
```
You can do `use copypasta::{Clipboard, Load, Buffer as ClipboardBuffer};` here.

cathay4t commented:
New patch uploaded. Changes since last one:

 * Will copy to both the selection and primary clipboards if `save_to_clipboard: true`.
 * `CHANGELOG.md` updated to include this changes.
 * Rephrased the comments of config file.

chrisduerr commented on src/input.rs:
```diff
@@ -469,7 +470,12 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
+            self.ctx.copy_selection(ClipboardBuffer::Selection);
+        } else {
+            self.ctx.copy_selection(ClipboardBuffer::Selection);
+        }
```
This code could be refactored to the following:

```rust
if self.save_to_clipboard {
    self.ctx.copy_selection(ClipboardBuffer::Primary);
}
self.ctx.copy_selection(ClipboardBuffer::Selection);
```

cathay4t commented:
Thanks for the review. New patch updated.

cathay4t commented:
Thanks for the review. Rephrased.
